File: src/main/scala/edu/uncc/cs/bridges/Bridge.scala
package edu.uncc.cs.bridges
import java.util
import java.util.ArrayDeque
import org.json.simple._
import scala.collection._
import scala.collection.JavaConverters._
import scala.collection.mutable.Set
import scala.collection.mutable.ListBuffer

/**
 * Network-enabled sample data aggregator.
 * Bridges offers connectivity for students to more easily use interesting real
 * world data for introductory projects.
 */
class Bridge(val assignment: Int) extends KeyConnectable {
    
    /** 
     * Connect to a streaming data source such as a social network feed.
     * This feature is not yet complete. Use at your own peril.
     * 
     * @unimplemented
     * @param name  The name of the stream to receive
     */
    def stream(name: String)= {
        new BStream(this, name)
    }
    
    /**
     * Upload visualization data.
     * Note that only one contiguous group will be shown.
     * So if you only see half of the graph you expected, make sure that there
     * is a path between the center node and the missing nodes.
     *  
     *  @param graph: a Graph or a subclass of Graph
     *  @param center: the central node of the visualization
     *  
     *  @return JSON-encoded String
     */
    def displayGraph(center: String, graph: Graph)= {
    	// Create and upload JSON
        post("/assignments/$assignment", generateGraphJson(center, graph))
        // Return a URL to the user
        s"/assignments/${assignment}/YOUR_USERNAME"
    }
    
    /**
     * Split an identifier into its constituent parts.
     * An identifier should look like "example.com/username".
     * @private
     * @param identifier
     * @return a length-2 String Array of [provider, username]
     */
    def splitIdentifier(identifier: String, provider_optional: Boolean=false)= {
    	if (! identifier.contains("/")) {
    		if (provider_optional)
    			Array("", identifier)
    		else
    			throw new RuntimeException(s"Provider or screenname missing in $identifier. Should look like: example.com/username")
    	} else {
    		identifier.split("/", 2)
    	}
    	 
    }
    
    /**
     * Generate the JSON to be uploaded.
     * It uses a BFS, and expects one contiguous group of nodes.
     * 
     * So if you only see half of the graph you expected, make sure that there
     * is a path between the center node and the missing nodes.
     * 
     *  This method is not idempotent; it will edit the graph and will not
     *  produce an accurate graph the second time.
     *  
     *  This method can be useful for debugging. But you should usually prefer
     *  to use displayGraph.
     *  
     *  @param graph: a Graph or a subclass of Graph
     *  @param center: the central node of the visualization
     * 
     */
    def generateGraphJson(center: String, graph: Graph)= {
    	val VISITED=1
    	val open = new ArrayDeque[String]();
    	graph.clearMarks();
    	graph.setMark(center, VISITED);
    	open.add(center);
    	
    	// These maps will be JSON objects
    	var edges = ListBuffer[Map[String, String]]()
    	var nodes = ListBuffer[Map[String, String]]()
    	// Gives the nodes a natural order
    	var names = ListBuffer[String]()
    	  
    	// Traverse the graph, as a BFS (although a DFS would work well too) 
    	while (! open.isEmpty()) {
    	  val local = open.remove()
    	  val name = splitIdentifier(local, true)(1)
    	  nodes += Map(
    	      "name" -> name,
    	      "color" -> Option(graph.getNodeColor(local)).getOrElse("")
	      )
	      names += name
    	  var neighbor = graph.first(local)
    	  while (neighbor != null) {
    	      edges += Map(
    	          "source" -> name,
    	          "target" -> splitIdentifier(neighbor, true)(1),
    	          "color" -> Option(graph.getEdgeColor(local, neighbor)).getOrElse(""),
    	          "value" -> graph.getEdge(local, neighbor).toString()
	          )
    	      if (graph.getMark(neighbor) != VISITED) {
    	    	  graph.setMark(neighbor, VISITED)
    	    	  open.add(neighbor);
    	      }
    	      neighbor = graph.next(local)
    	  }
    	}
    	
    	// Convert the JSON-like maps into real JSON
    	var node_string = nodes
			.toList.sortBy( _("name") )  // Improve testability
			.map {
    			_
    				.map({case (k, v) => s""""$k":"$v""""})
    				.reduceOption(_+","+_)
    				.getOrElse("")
    		}
			.map(a => s"{$a}")
			.reduceOption(_+","+_)
			.map(a => s"[$a]")
			.getOrElse("[]")
		
		// Repeat for edges
		names = names.sorted
    	var edge_string = edges
			.map { m =>
				s"""{"source":${names.indexOf(m("source"))},"target":${names.indexOf(m("target"))},"color":"${m("color")}","value":${m("value")}}"""
			}
			.reduceOption(_+","+_)
			.map(a => s"[$a]")
			.getOrElse("[]")
    	s"""{"nodes":$node_string,"links":$edge_string}"""
    }
    
    /** Automatically choose whether to open a node identifier with:
     *  Twitter via followers(),
     *  TMDb with movies(), or
     *  RottenTomatoes with actors()
     *  
     *  @returns an array of identifiers
     */
    def neighbors(identifier: String, max: Integer=100)= {
    	val Array(service, screen_name) = splitIdentifier(identifier) 
    	if (service.equals("twitter.com"))
    		followers(identifier, max)
		else if (service.equals("actor"))
			movies(identifier, max)
		else if (service.equals("movie"))
			actors(identifier, max)
		else
		    //List[String]().asJava
			throw new RuntimeException(s"Unrecognized service $service. Choose from twitter.com, actor, or movie")
    }
    
    /** List the user's followers as more FollowGraphNodes.
        Limit the result to `max` followers. Note that results are batched, so
        a large `max` (as high as 200) _may_ only count as one request.
        See Bridges.followgraph() for more about rate limiting. */
    def followers(screen_name_identifier: String, max: Integer=100)= {
    	val Array(service, screen_name) = splitIdentifier(screen_name_identifier, true)
        (try {
	        val response = getjs(s"/streams/twitter.com/followers/$screen_name/$max")
	        val users = response.get("followers").asInstanceOf[util.List[JSONValue]]
	        if (users == null) {
	            // Twitter sometimes gives us garbage
	    		List()
	        } else {
	        	(0 until users.size()).map {
	        		i => s"twitter.com/${users.get(i).asInstanceOf[String]}"
	        	}
	        }
        } catch {
        	case e: RateLimitException => List()
    	}).asJava
    }
    
    /**
     * Return a list of movies an actor played in.
     * 
     * The data comes courtesy of RottenTomatoes.
     * 
     * The quota for this resource is about 10k actors/day but is shared by all
     * students. So if you consume all 10k, it will be a bad day. Please make
     * sure you limit your queries appropriately.
     * 
     */
    def movies(actor_identifier: String, max: Integer=100)= {
    	val Array(service, actor) = splitIdentifier(actor_identifier, true) 
        (try {
	        val response = getjsarray(s"/streams/actors/$actor")
	        val movies_json = response.asInstanceOf[util.List[JSONValue]]
	        // Get (in JS) movies_json.map(function(m) { return m.title; }) 
        	(0 until movies_json.size()).take(max).map {
        		i => movies_json
    				.get(i).asInstanceOf[JSONObject]
    				.get("title").asInstanceOf[String]
        	}.map { "movie/" + _ }
        } catch {
        	case e: RateLimitException => List()
        	// TODO: Decide whether to warn, error, or eat cast exceptions
        	case e: ClassCastException => List()
    	}).asJava
    }
    
    /**
     * Return the actors that played in a movie.
     * 
     * The data comes courtesy of TMDb.
     * 
     * This resource has unlimited queries but has caveats. Not every extra
     * that played in every movie ever is listed in the database and some
     * movies are documented rather sparsely. Expect some to be missing.
     */
    def actors(movie_identifier: String, max: Integer=100)= {
    	val Array(service, movie) = splitIdentifier(movie_identifier, true) 
        (try {
	        val response = getjsarray(s"/streams/rottentomatoes.com/$movie")
	        // We will assume that the first movie is the right one
	        val actors_json = response.asInstanceOf[util.List[JSONValue]]
	        if (actors_json.isEmpty()) {
	        	// There wasn't a first movie.
	        	List()
        	} else {
        		// Get (in JS) movies[0].abridged_cast.map(function(a) {return a.name;})
	        	val abridged_cast = actors_json
	        		.get(0).asInstanceOf[JSONObject]
    				.get("abridged_cast").asInstanceOf[JSONArray];
	        	(0 until abridged_cast.size()).take(max).map {
	        		i => abridged_cast
	    				.get(i).asInstanceOf[JSONObject]
	    				.get("name").asInstanceOf[String]
	        	}.map { "actor/" + _ }
        	}
        } catch {
        	case e: RateLimitException => List()
        	// TODO: Decide whether to warn, error, or eat cast exceptions
        	case e: ClassCastException => List()
    	}).asJava
    }
}

File: src/main/java/edu/uncc/cs/bridges/Bridge.java
package edu.uncc.cs.bridges;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

import org.json.simple.JSONArray;
import org.json.simple.JSONObject;


/**
 * Connection to the Bridges server.
 * 
 * Initialize this class before using it, and call complete() afterward.
 * 
 * @author Sean Gallagher
 */
public class Bridge {
	private static int assignment;
	private static String key;
	private static Visualizer visualizer;
	private static BridgeNetwork backend;
	private static String userName;	
	
	// This exists to prevent duplicate error traces.
	public static String getUserName() {
		return userName;
	}

	public static void setUserName(String userName) {
		Bridge.userName = userName;
	}

	private static boolean failsafe = false;

	/**
	 * Initialize Bridges with Visualizer
	 * @param assignment  The assignment number, for grading
	 * @param visualizer  The visualizer, for assignment
	 * @param username TODO
	 */
	public static void init(int assignment, String key, Visualizer visualizer, String username) {
		Bridge.assignment = assignment;
		Bridge.key = key;
		Bridge.visualizer = visualizer;
		Bridge.backend = new BridgeNetwork();
		Bridge.userName = username;
	}
	
	/* Accessors and Mutators */

	public static int getAssignment() {
		return assignment;
	}
	public static void setAssignment(int assignment) {
		Bridge.assignment = assignment;
	}
	
	public static String getKey() {
		return key;
	}
	public static void setKey(String key) {
		Bridge.key = key;
	}

	public static String getServerURL() {
		return backend.server_url;
	}
	public static void setServerURL(String server_url) {
		Bridge.backend.server_url = server_url;
	}

	public static Visualizer getVisualizer() {
		return visualizer;
	}
	public static void setVisualizer(Visualizer visualizer) {
		Bridge.visualizer = visualizer;
	}
	
	/**
	 * Update visualization metadata. This may be called many times.
	 * This is usually an expensive operation and involves connecting to the network.
	 * Calling this method is optional provided you call complete()
	 */
	public static void update() {
        try {
        	//System.out.println(visualizer.getRepresentation());
			backend.post("/assignments/" + assignment, visualizer.getRepresentation());
		} catch (IOException e) {
			System.err.println("There was a problem sending the visualization"
					+ " representation to the server. Are you connected to the"
					+ " Internet? Check your network settings. Otherwise, maybe"
					+ " the central Bridges server is down. Try again later.\n"
					+ e.getMessage());
		} catch (RateLimitException e) {
			System.err.println("There was a problem sending the visualization"
					+ " representation to the server. However, it responded with"
					+ " an impossible 'RateLimitException'. Please contact"
					+ " Bridges developers and file a bug report; this error"
					+ " should not be possible.\n"
					+ e.getMessage());
		} 
        // Return a URL to the user
        System.out.println("Check out your visuals at " + backend.prepare("/assignments/" + assignment + "/" + userName) );
	}

	/**
	 * Close down Bridges.
	 * 
	 * This only calls update() but it could conceivably do more.
	 */
	public static void complete() {
		update();
	}
	
	
	// Internal utility methods
	
	/**
	 * Internal method for JSON preparation
	 * @param in 	The original string
	 * @return a string with all but the last character
	 */
	public static StringBuilder trimComma(StringBuilder in) {
		if (in.length() > 0 && in.charAt(in.length()-1) == ',')
			in.deleteCharAt(in.length()-1);
		return in;
	}
	
	/**
	 * Internal method for JSON preparation
	 * @param in	The original String
	 * @return the string, encapsulated in quotes
	 */
	static String quote(String in) {
		return String.format("\"%s\"", in);
	}
	
	/**
	 * Internal method for JSON preparation
	 * @return a string with all but the first and last characters
	 */
	static String unquote(String in) {
		return in.substring(
				Math.min(in.length()-1, 1),
				Math.max(in.length()-1, 0));
	}
	
	/**
	 * Idiom for enabling ordered iteration on any map.
	 * The reason for this is to make the strings compare equal for testing
	 * @param values
	 * @return
	 */
	static <T extends Comparable<T>> List<T> sorted_values(
			Map<String, T> values) {
		List<T> sorted_values = new ArrayList<>(values.values());
		Collections.sort(sorted_values);
		return sorted_values;
	}
	
	/**
	 * Idiom for enabling ordered iteration on any map.
	 * The reason for this is to make the strings compare equal for testing
	 * @param values
	 * @return
	 */
	static <K extends Comparable<K>, V> List<Entry<K, V>> sorted_entries(
			Map<K, V> map) {
		List<Entry<K, V>> sorted_entries = new ArrayList<>(map.entrySet());
		Collections.sort(sorted_entries, new Comparator<Entry<K, V>>() {
			public int compare(Entry<K, V> t, Entry<K, V> o) {
				return t.getKey().compareTo(o.getKey());
			}
		});
		return sorted_entries;
	}
	
    /**
     * 	Automatically choose whether to open a node identifier with:
     *  Twitter via followers(),
     *  TMDb with movies(), or
     *  RottenTomatoes with actors()
     * 
     * @param identifier
     * @param max
     * @throws IOException 
     * @returns a list of identifiers
     * @throws QueryLimitException
     */
    public static List<String> getAssociations(String identifier, int max){
    	try {
	    	Ident id = Ident.fromAnyString(identifier);
	    	switch (id.provider) {
	    	case "twitter.com":
	    		return followers(id, max);
	    	case "actor":
	    		return movies(id, max);
	    	case "movie":
	    		return actors(id, max);
	    	default:
	    		throw new RuntimeException(
	    				"Unrecognized service " + id.provider
	    				+ ". Choose from twitter.com, actor, or movie");
	    	}
    	} catch (RateLimitException e) {
    		return new ArrayList<>();
    	}
    	
    }
    
    /** List the user's followers as more FollowGraphNodes.
        Limit the result to `max` followers. Note that results are batched, so
        a large `max` (as high as 200) _may_ only count as one request.
        See Bridges.followgraph() for more about rate limiting. 
     * @throws IOException */
    static List<String> followers(Ident id, int max)
    		throws RateLimitException {
    	if (failsafe) {
    		// Don't contact Twitter, use sample data
    		return SampleDataGenerator.getFriends(id.name, max);
    	} else {
	    	try {
		    	String resp = backend.get("/streams/twitter.com/followers/"
		    			+ id.name + "/" + max);
		    	
		        JSONObject response = backend.asJSONObject(resp);
		        JSONArray followers = (JSONArray) backend.safeJSONTraverse(
		        		"['followers']", response, JSONArray.class);
		        
		        List<String> results = new ArrayList<>();
		    	for (Object follower : followers) {
		    		String name = (String) backend.safeJSONTraverse(
		    				"", follower, String.class);
		    		results.add("twitter.com/" + name);
		    	}
		    	return results;
	    	} catch (IOException e) {
	    		// Trigger failsafe.
	    		System.err.println("Warning: Trouble contacting Bridges. Using "
	    				+ "sample data instead.\n"
	    				+ e.getMessage());
	    		failsafe = true;
	    		return followers(id, max);
	    	}
    	}
    }
    
    /**
     * Return a list of movies an actor played in.
     * 
     * The data comes courtesy of RottenTomatoes.
     * 
     * The quota for this resource is about 10k actors/day but is shared by all
     * students. So if you consume all 10k, it will be a bad day. Please make
     * sure you limit your queries appropriately.
     * 
     */
    static List<String> movies(Ident id, int max)
    		throws RateLimitException {

    	if (failsafe) {
    		// Don't contact Bridges, use sample data
    		return SampleDataGenerator.getFriends(id.name, max);
    	} else {
	    	try {
		    	String resp = backend.get("/streams/actors/" + id.name);
		    	JSONArray movies = backend.asJSONArray(resp);
		    	
		        // Get (in JS) movies_json.map(function(m) { return m.title; })
		        List<String> results = new ArrayList<>();
		        for (Object movie : movies) {
		        	String title = (String) backend.safeJSONTraverse("['title']",
		        			movie, String.class);
		        	results.add("movie/" + title);
		        }
		        return results;
	    	} catch (IOException e) {
	    		// Trigger failsafe.
	    		System.err.println("Warning: Trouble contacting Bridges. Using "
	    				+ "sample data instead.\n"
	    				+ e.getMessage());
	    		failsafe = true;
	    		return movies(id, max);
	    	}
    	}
    }
    
    /**
     * Return the actors that played in a movie.
     * 
     * The data comes courtesy of TMDb.
     * 
     * This resource has unlimited queries but has caveats. Not every extra
     * that played in every movie ever is listed in the database and some
     * movies are documented rather sparsely. Expect some to be missing.
     * @throws IOException 
     * @throws RateLimitException 
     */
    static List<String> actors(Ident id, int max)
    		throws RateLimitException {

    	if (failsafe) {
    		// Don't contact Bridges, use sample data
    		return SampleDataGenerator.getFriends(id.name, max);
    	} else {
	    	try {
		    	String resp = backend.get("/streams/rottentomatoes.com/" + id.name);
		    	JSONArray movies = backend.asJSONArray(resp);
		    	
		        // We will assume that the first movie is the right one
		    	JSONArray abridged_cast = (JSONArray) backend.safeJSONTraverse(
		    			"[0]['abridged_cast']", movies, JSONArray.class);
		    	List<String> results = new ArrayList<>();
		    	for (Object cast_member : abridged_cast) {
		    		if (results.size() == max)
		    			break;
		    		String name = (String) backend.safeJSONTraverse("['name']",
		    				cast_member, String.class);
					results.add("actor/" + name);
		    	}
		    	return results;
	    	} catch (IOException e) {
	    		// Trigger failsafe.
	    		System.err.println("Warning: Trouble contacting Bridges. Using "
	    				+ "sample data instead.\n"
	    				+ e.getMessage());
	    		failsafe = true;
	    		return actors(id, max);
	    	}
    	}
    }
    
    /**
     * Generate a sample Edge weight for two nodes
     * @param source
     * @param target
     * @return
     */
    public static double getEdgeWeight(String source, String target) {
    	int h = source.hashCode() + target.hashCode();
    	if (h < 0) h = -h;
    	return h % 10;
    }
}

class Ident {
	public String provider;
	public String name;
	
	/**
	 * Create a new Ident from an identifier string with a provider
	 * @param identifier
	 */
	public Ident(String identifier) {
    	if (identifier.contains("/")) {
    		String[] halves = identifier.split("/", 2);
    		provider = halves[0];
    		name = halves[1];
    	} else {
    		throw new RuntimeException("Provider or screenname missing in "
    				+ identifier + ". Should look like: example.com/username");
    	}
	}
	
	/**
	 * Create a plain, straightforward Ident with the provider and name as given.
	 * @param provider
	 * @param name
	 */
	public Ident(String provider, String name) {
		this.provider = provider;
		this.name = name;
	}
	
	public static Ident fromAnyString(String identifier) {
    	if (identifier.contains("/")) {
    		return new Ident(identifier);
    	} else {
    		return new Ident("", identifier);
    	}
	}
}
